### PR TITLE
fix!: Make for loops a statement

### PR DESCRIPTION
**Pull request by jfecher**
# Description

## Problem\*

Resolves 

## Summary\*

This fixes another issue hiding in #2911. Previously there was a type checker bug. This PR fixes the parse of `for ... {} *foo` to parse `for .. {}; *foo` rather than multiplying the result of the for with foo.

Changes for loops to be parsed as statements rather than expressions. Due to how chumsky works with needing to hoist up recursive parsers into parameters, this required many more changes than I would have liked.

This is breaking since code like `let x = for i in 0..1 {};` no longer compiles. I don't think code like this is used in practice however.

## Documentation

- [ ] This PR requires documentation updates when merged.

  

  - [ ] I will submit a noir-lang/docs PR.

  

  - [ ] I will request for and support Dev Rel's help in documenting this PR.

  
  

## Additional Context

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Review by kevaundray — APPROVED** [review]
This was here for historical reasons;

When we did not have a concept of mutability, this way was used to modify an array :D

File: compiler/noirc_frontend/src/parser/parser.rs [hunk 1]
```diff
@@ -1343,25 +1366,21 @@ where
...
+fn if_expr<'a, P, S>(expr_no_constructors: P, statement: S) -> impl NoirParser<ExpressionKind> + 'a
 where
-    P1: ExprParser + 'a,
-    P2: ExprParser + 'a,
+    P: ExprParser + 'a,
+    S: NoirParser<Statement> + 'a,
 {
     recursive(|if_parser| {
-        let if_block = block_expr(expr_parser.clone());
+        let if_block = block_expr(statement.clone());
```
Comment (kevaundray) [inline compiler/noirc_frontend/src/parser/parser.rs:1375]: Oh didn't see this -- why was this changed?

File: compiler/noirc_frontend/src/hir/resolution/resolver.rs [hunk 2]
```diff
@@ -957,6 +957,25 @@ impl<'a> Resolver<'a> {
                 let stmt = HirAssignStatement { lvalue: identifier, expression };
                 HirStatement::Assign(stmt)
             }
+            Statement::For(for_loop) => {
+                let start_range = self.resolve_expression(for_loop.start_range);
+                let end_range = self.resolve_expression(for_loop.end_range);
+                let (identifier, block) = (for_loop.identifier, for_loop.block);
+
+                // TODO: For loop variables are currently mutable by default since we haven't
```
Comment (vezenovm) [inline compiler/noirc_frontend/src/hir/resolution/resolver.rs:965]: Non-blocking but do we have an issue for this?

Comment (vezenovm) [reply compiler/noirc_frontend/src/hir/resolution/resolver.rs:965]: Ah I see this comment is leftover

Comment (jfecher) [reply compiler/noirc_frontend/src/hir/resolution/resolver.rs:965]: I don't believe we do

File: compiler/noirc_frontend/src/parser/parser.rs (hunk 1 above)
Comment (jfecher) [reply compiler/noirc_frontend/src/parser/parser.rs:1375]: The `statement` parser is now recursive. It was surprising it wasn't recursive until this point. For loops are the only statement type now to have blocks in them. The others only have expressions with blocks - for which they can just use the already recursive expression parser.

The fact that the statement parser is now recursive is why I had to change so many other parsers just to pass it around now. This is my main annoyance with chumsky.

Comment (kevaundray) [reply compiler/noirc_frontend/src/parser/parser.rs:1375]: oof

**Review by kevaundray — APPROVED** [review]

**Review by kevaundray — APPROVED** [review]
